### DISCUSSION TITLE
How to set custom root (endpoint_url) for corporate aws

### DISCUSSION THREAD

**Discussion by asant-leitha** [aws/aws-toolkit-vscode#3960] [Q&A]
In order to access my corporate aws instance I need to set a custom root otherwise I get 403 Unauthorized. How to do that using the config file or the vscode toolkit plugin?

**Comment by richardhboyd** [comment]
Can you share a little more context about your setup? Are you referring to VSCode's [multi-root workspace](https://code.visualstudio.com/docs/editor/multi-root-workspaces)? What action is returning the 403 error?

**Reply by asant-leitha** [reply]
Ofc. This is my config file

```
[default]
endpoint_url = https://url.it:port
aws_access_key_id = XXXXXXXXXX
aws_secret_access_key = xxxxxxxx/xxxxxxx
ca-bundle = "path/to/file.pem"
```

When I try to connect using the AWS Explorer
![image](https://github.com/aws/aws-toolkit-vscode/assets/149064657/6cd57d17-0a15-4288-9bb5-000dc1a1b357)

I get the following error

```
2023-10-26 14:33:02 [ERROR]: _aws.auth.reauthenticate: Error: Unable to authenticate connection
	 -> InvalidClientTokenId: The security token included in the request is invalid. (statusCode: 403; requestId: f606ee3a-63b3-4085-8671-5742a15534b4)
```

**Reply by richardhboyd** [reply]
Can you run `aws sts get-caller-identity` from your command line to ensure that your credentials are valid?

**Reply by asant-leitha** [reply]
They are valid as I can successfully authenticate using DataGrip. The only difference is that I need to set a custom root to get access

**Reply by richardhboyd** [reply]
I'm still unclear what the custom root is. are you referring to the root of the workspace? where are you setting that?

**Reply by asant-leitha** [reply]
![image](https://github.com/aws/aws-toolkit-vscode/assets/149064657/e7fd0864-5e45-4c82-b198-50089fc015dd)
custom root is the red line

**Reply by richardhboyd** [reply]
let me check with the team that owns the IDE Extension.

**Comment by justinmk3** [comment]
Unless https://github.com/aws/aws-sdk-js-v3 implicitly supports `endpoint_url` , AWS Toolkit for VSCode has no explicit support for this. AWS Toolkit depends on the SDK to understand the credentials configuration. 

The only sdk issues I could find that might be related:
- https://github.com/aws/aws-sdk-js-v3/discussions/5189
- https://github.com/aws/aws-sdk-js-v3/pull/5312

**Comment by justinmk3** [comment]
I created https://github.com/aws/aws-toolkit-vscode/issues/3961 to track this as a feature request.

> This is my config file
> 
> ```
> [default]
> endpoint_url = https://url.it:port
> aws_access_key_id = XXXXXXXXXX
> aws_secret_access_key = xxxxxxxx/xxxxxxx
> ca-bundle = "path/to/file.pem"
> ```

Does that config work with AWS CLI?

- `ca-bundle` is not listed on https://awscli.amazonaws.com/v2/documentation/api/latest/topic/config-vars.html . Did you mean `ca_bundle` ?
- `endpoint_url` is documented at https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-endpoints.html but not sure if the SDK supports it.

**Reply by asant-leitha** [reply]
changing to `ca_bundle` won't help

**Reply by asant-leitha** [reply]
Tried to run

```
aws s3 ls
```

but I get
```
SSL validation failed for https://url.it:port/ [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate (_ssl.c:1006)
```

**Reply by asant-leitha** [reply]
Running it using the option `--no-verify-ssl` will print

```
An error occurred (AccessDenied) when calling the ListBuckets operation: Access Denied
```

because I do not have access to all roots

**Reply by justinmk3** [reply]
"Roots" appears to be a DataGrip concept that I am not familiar with. Is there documentation for it?

**Reply by asant-leitha** [reply]
Basically it is a root path in fact I am successfully connecting to my corporate aws instance by launching

```
aws s3 ls s3://custom-root --no-verify-ssl
```

**Reply by justinmk3** [reply]
Ok so "root" is equivalent to `endpoint_url`?

**Reply by asant-leitha** [reply]
No the endpoint is the url of the corporate aws instance and the root is the root path of the buckets
